source/shared/core_conn.cpp

```diff
@@ -150,34 +150,134 @@ sqlsrv_conn* core_sqlsrv_connect( _In_ sqlsrv_context& henv_cp, _In_ sqlsrv_cont
-        r = core_odbc_connect( conn, conn_str, is_missing_driver, is_pooled );
+            // check if ODBC 17 actually exists, if not, throw an exception
+            CHECK_CUSTOM_ERROR( ! core_search_odbc_driver_unix( ODBC_DRIVER_17 ), conn, SQLSRV_ERROR_SPECIFIED_DRIVER_NOT_FOUND ) {
+                throw core::CoreException();
+            }
 
-        CHECK_CUSTOM_ERROR( is_missing_driver, conn, SQLSRV_ERROR_AE_DRIVER_NOT_INSTALLED, get_processor_arch()) {
-            throw core::CoreException();
+            r = core_odbc_connect( conn, conn_str, is_pooled );
         }
+        else {  
+            // driver not specified, so connect using ODBC 17
+            // In non-Windows environment, unixODBC 2.3.4 and unixODBC 2.3.1 return different error states 
+            // If it fails to connect using ODBC 17, it is unreliable to check for a certain sql state error.  
+            // Thus, we will simply throw an exception.
+            if( core_search_odbc_driver_unix( ODBC_DRIVER_17 ) ) {
+                conn_str = conn_str + CONNECTION_STRING_DRIVER_NAME[ DRIVER_VERSION::ODBC_DRIVER_17 ];
+                r = core_odbc_connect( conn, conn_str, is_pooled );
+            }
+            else {
+                CHECK_CUSTOM_ERROR( true, conn, SQLSRV_ERROR_AE_DRIVER_REQUIRED, get_processor_arch()) {
+                    throw core::CoreException();
+                }
+            }
+        } // else driver_version not unknown
     }
-    else {
+    else { 
+        // column encryption NOT enabled
+        if( conn->driver_version != ODBC_DRIVER_UNKNOWN ) {
+            // check if the ODBC driver actually exists, if not, throw an exception
+            CHECK_CUSTOM_ERROR( ! core_search_odbc_driver_unix( conn->driver_version ), conn, SQLSRV_ERROR_SPECIFIED_DRIVER_NOT_FOUND ) {
+                throw core::CoreException();
+            }
 
-        for ( std::size_t i = DRIVER_VERSION::FIRST; i <= DRIVER_VERSION::LAST; ++i ) {
-            is_missing_driver = false;
-            std::string conn_str_driver = conn_str + CONNECTION_STRING_DRIVER_NAME[ DRIVER_VERSION(i) ];
-            r = core_odbc_connect( conn, conn_str_driver, is_missing_driver, is_pooled );
-            CHECK_CUSTOM_ERROR( is_missing_driver && ( i == DRIVER_VERSION::LAST ), conn, SQLSRV_ERROR_DRIVER_NOT_INSTALLED, get_processor_arch()) {
+            r = core_odbc_connect( conn, conn_str, is_pooled );
+        }
+        else {
+            DRIVER_VERSION odbc_version = ODBC_DRIVER_UNKNOWN;
+            for( short i = DRIVER_VERSION::FIRST; i <= DRIVER_VERSION::LAST; ++i ) {
+                // skip ODBC 11 in a non-Windows environment -- only available in Red Hat / SUSE (preview)
+                // https://docs.microsoft.com/en-us/sql/connect/odbc/linux-mac/installing-the-microsoft-odbc-driver-for-sql-server#microsoft-odbc-driver-11-for-sql-server-on-linux
+                if (i == DRIVER_VERSION::ODBC_DRIVER_11)    
+                    continue;
```

Comment:
I come from the camp that considers continue, break, goto as bad programming style.  In this case it can be easily avoided by adding this condition in the next if